persistence/block.go

```diff
@@ -76,64 +73,86 @@ func (p PostgresContext) TransactionExists(transactionHash string) (bool, error)
-//                 until we include the schema as part of the SQL Store because persistence
-//                 currently has no access to the protobuf schema which is the source of truth.
-// TODO: atomic operations needed here - inherited pattern from consensus module
-func (p PostgresContext) storeBlock(quorumCert []byte) error {
-	if p.blockProtoBytes == nil {
-		// IMPROVE/CLEANUP: HACK - currently tests call Commit() on the same height and it throws a
-		// ERROR: duplicate key value violates unique constraint "block_pkey", because it attempts to
-		// store a block at height 0 for each test. We need a cleanup function to clear the block table
-		// each iteration
-		return nil
+// Creates a block protobuf object using the schema defined in the persistence module
+func (p *PostgresContext) prepareBlock(quorumCert []byte) (*types.Block, error) {
+	var prevHash []byte
+	if p.Height == 0 {
+		prevHash = []byte("")
+	} else {
+		var err error
+		prevHash, err = p.GetBlockHash(p.Height - 1)
+		if err != nil {
+			return nil, err
+		}
 	}
-	// INVESTIGATE: Note that we are writing this directly to the blockStore. Depending on how
-	// the use of the PostgresContext evolves, we may need to write this to `ContextStore` and copy
-	// over to `BlockStore` when the block is committed.
-	if err := p.blockstore.Put(heightToBytes(p.Height), p.blockProtoBytes); err != nil {
-		return err
+
+	txsHash, err := p.getTxsHash(p.Height)
+	if err != nil {
+		return nil, err
 	}
-	// Store in SQL Store
-	if err := p.InsertBlock(uint64(p.Height), p.blockHash, p.proposerAddr, quorumCert); err != nil {
-		return err
+
+	block := &types.Block{
+		Height:            uint64(p.Height),
+		Hash:              p.blockHash,
+		PrevHash:          hex.EncodeToString(prevHash),
+		ProposerAddress:   p.proposerAddr,
+		QuorumCertificate: quorumCert,
+		TransactionsHash:  txsHash,
 	}
-	// Store transactions in indexer
-	return p.indexTransactions()
+
+	return block, nil
 }
 
-func (p PostgresContext) InsertBlock(height uint64, hash string, proposerAddr []byte, quorumCert []byte) error {
+// Inserts the block into the postgres database
+func (p *PostgresContext) insertBlock(block *types.Block) error {
 	ctx, tx, err := p.GetCtxAndTx()
 	if err != nil {
 		return err
 	}
 
-	_, err = tx.Exec(ctx, types.InsertBlockQuery(height, hash, proposerAddr, quorumCert))
+	_, err = tx.Exec(ctx, types.InsertBlockQuery(block.Height, block.Hash, block.ProposerAddress, block.QuorumCertificate))
 	return err
 }
 
-// CLEANUP: Should this be moved to a shared directory?
+// Stores the block in the key-value store
+func (p PostgresContext) storeBlock(block *types.Block) error {
+	blockBz, err := codec.GetCodec().Marshal(block)
+	if err != nil {
+		return err
+	}
+	return p.blockStore.Set(heightToBytes(p.Height), blockBz)
+}
+
+func (p PostgresContext) getTxsHash(height int64) ([]byte, error) {
+	// The order (descending) is important here since it is used to comprise the hash in the block
```

Comment:
Either the comment or implementation is wrong:

<img width="748" alt="Screen Shot 2022-11-09 at 5 22 01 PM" src="https://user-images.githubusercontent.com/18633790/200954885-3a3aa1c0-ce75-416e-a844-9cc0acf71899.png">